lib/slack/real_time/socket.rb

```diff
@@ -30,9 +35,46 @@ def connect!
         connect
         logger.debug("#{self.class}##{__method__}") { driver.class }
 
+        heartbeat 30
         yield driver if block_given?
       end
 
+      def heartbeat(delay)
+        @alive = true
+
+        driver.on :message do |event|
+          event_data = JSON.parse(event.data)
+          pong(event_data['reply_to']) if event_data['type'] == 'pong'
+        end
+
+        Thread.new do
+          ping_int = 0
+          loop do
+            sleep delay
+            ping_data = { type: 'ping', time: Time.now, id: "p#{ping_int += 1}" }
+            ping(ping_data)
+          end
+        end
```

Comment:
Where do you call `Thread#join`? If two threads try to write to the socket at the same time, what happens when the data becomes interleaved?

Reply:
No worries. Please do make suggestions about this implementation! There's a lot of looping around in that calling code :)